tests/speed_sig_stfl.c

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: MIT
+
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <oqs/oqs.h>
+
+#if defined(OQS_USE_RASPBERRY_PI)
+#define _OQS_RASPBERRY_PI
+#endif
+#if defined(OQS_SPEED_USE_ARM_PMU)
+#define SPEED_USE_ARM_PMU
+#endif
+#include "ds_benchmark.h"
+#include "system_info.c"
+
+OQS_STATUS dummy_secure_storage(uint8_t *sk_buf, size_t sk_buf_len, void *context) {
+	// suppress unused parameter warning
+	(void)(sk_buf);
+	(void)(sk_buf_len);
+	(void)(context);
+	return OQS_SUCCESS;
+}
+
+// reset secret key: some schemes fail to create a new secret key over a previous secret key
```

Comment:
I'm curious about this comment: I would have expected all the schemes to behave similarly. Do you have an idea as to why this occurs (or at least which schemes behave differently)?

Reply:
Thanks for the detailed response! Looping in @ashman-p and @cothan as the primary authors of the LMS / XMSS code to see if this is something we ought to fix.

Reply:
Note: I actually disagree with not accepting context=NULL.
It might be a valid choice to use context to pass the index of the relevant secret key, which could start at 0.
Or have a secure store for exactly 1 key, and leave context NULL.

Reply:
Hi @cr-marcstevens  , thanks for the review of the code. I like your propose. 
I think we can reiterate the code better. Since we merge the LMS and XMSS, there are not many issues opened to improve the interface (perhaps no one use it yet?), so this can be one. 
We can merge this PR in first and solve this problem in separately. What do you think?

Reply:
That seems like a good choice to first merge this PR, otherwise it will likely go stale, while it is of direct use for benchmarking (at least for us) and comparing schemes.

However it seems it currently doesn't pass CI on a couple of tests.
Some of these simply timeout after 60 minutes, so perhaps sig stfl tests should be quite limited as these can take really long.

Other tests I'm not sure what the problem is.

There is one issue that I'm also aware of:
There is a memory leak when freeing the secret key, because the OQS_SIG_STFL_SECRET_KEY_free functions only clears and frees up the internal structure, but not the struct that secret key itself point to.
See:
https://github.com/open-quantum-safe/liboqs/blob/81b4452b9a653f5e203b4bd48ff6fa87a65d0147/src/sig_stfl/xmss/sig_stfl_xmss_secret_key_functions.c#L137

Since speed_sig_stfl needs to free and new secret keys often, this might be  a particular issue in running these benchmarks.

Reply:
@cr-marcstevens, echoing gratitude for this contribution and feedback. As you pointed out there is need to protect the secret key. Protect it from reuse and loss. As a result all that weight was passed on to the caller of the API.
Perhaps better documentation would be helpful. 

Similarly, AFAIR, i think caller is also tasked with freeing memory foe the actual secret key (did not want the library to blindly free a void pointer).  
This feature was perhaps one of the more contested ones hence the hoops needed to enable key and signature generation.

Reply:
In this case the user needs to create a secret pointer through SECRET_KEY_new.
This functions first mallocs a general OQS_SIG_STFL_SECRET_KEY object (to which a pointer is returned in the end), and then mallocs a scheme dependent internal secret_key_data.
See:
https://github.com/open-quantum-safe/liboqs/blob/81b4452b9a653f5e203b4bd48ff6fa87a65d0147/src/sig_stfl/xmss/sig_stfl_xmss_secret_key_functions.c#L14

But the corresponding SECRET_KEY_free only frees the internal secret_key_data, but not the general object.
It needs another free after that to really clean things up.

I would argue against this, and have the general object be directly freed as well.
Because the naming suggests this paradigm anyway, but also because the partially freed object is of no further use: it cannot be reused, because any call to SECRET_KEY_NEW would malloc a new general object anyway, not reuse the current one.

Reply:
@cr-marcstevens , see the API call to free secret keys. 
[https://github.com/open-quantum-safe/liboqs/blob/81b4452b9a653f5e203b4bd48ff6fa87a65d0147/src/sig_stfl/sig_stfl.c#L1512](OQS_SIG_STFL_SECRET_KEY_free)

Reply:
@ashman-p Thanks for the pointer! I noticed this weird omission in the scheme specific code and thought there was a memory-leak, so I wanted to point that out.
Thanks for pointing out this is actually handled in the generic code, my mistake.

Reply:
Is this comment resolved? If so I can merge the PR.

Reply:
Thanks for pointing this out @SWilson4 ! @open-quantum-safe/liboqs-committers if you subscribe to the thought that [liboqs at some point in the future should be of "productive utility"](https://github.com/open-quantum-safe/tsc/issues/1) than please heed good practices contributing to eventually achieve this status: Existing tests must pass for PRs && new features must come with relevant new tests && PR discussions must be (marked) resolved (by the people raising the question). If you don't subscribe to this goal, well, then of course substantially less rigor can be applied, saving all of us time and effort and keep moving the library forward functionally as a research vehicle.

Reply:
@cr-marcstevens , oh right, I forget to mention that the Keygen operation should be skipped, since the Key generation can take from minutes to 3 hours to complete, (or only benchmark the keygen of the smallest parameters), so I think the benchmark probably should focus on Sign and Verify operations.

Reply:
So before merging can I ask what's the conclusion now? Skip keygen? Make this optional (such that it can be omitted from CI but can be manually executed)? Again, my recommendation is that the originator of the comment marks it resolved if he's content such as to assure someone like me to judge "closure" and merge.

Reply:
My current solution is to only test algorithms with 2^10 signatures.
That is all XMSS_*_10_* and LMS*_H10_*.
So these do get tested in CI, but the computational cost is quite low.

To that end `speed_sig_stfl` has an added option `--limit10`, which instead of all algs tests only those algs.

Reply:
BTW1 the current code requires keygen to generate secret keys, to then do signing and verification with. Avoiding keygen by eg serialization/deserialization of an offline generated secret, is I think it too much complication just for a benchmarking program.
BTW2 for LMS the secret key seems to be just the seed, and signing essentially requires recomputing the entire tree again anyway.

Reply:
Thanks for the explanations @cr-marcstevens . I now also saw the command line options to that effect . I'd now be OK merging this if @SWilson4 marks the conversation resolved.